polynote-frontend/polynote/toolbar.js

```diff
@@ -1,58 +1,253 @@
-import { FakeSelect } from './fake_select'
-
-export class Toolbar {
-    constructor(id, actions) {
-        this.element = (id instanceof HTMLElement) ? id : document.getElementById(id);
-        this.buttons = [];
-        for (const selector in actions) {
-            if (actions.hasOwnProperty(selector)) {
-                let action = actions[selector];
-                let button = this.element.querySelector(selector);
-
-                if (button.classList.contains('dropdown')) {
-                    button = new FakeSelect(button);
-                } else {
-                    button.setState = (state) => {
-                        if (state && state !== 'false') {
-                            button.classList.add('active');
+import {button, div, fakeSelectElem, h3, iconButton} from "./tags";
+import {FakeSelect} from "./fake_select";
+import {Cell, CodeCell, TextCell} from "./cell";
+import {LaTeXEditor} from "./latex_editor";
+import {UIEvent, UIEventTarget} from "./ui_event";
+
+export class ToolbarEvent extends UIEvent {
+    constructor(eventId, details) {
+        super(eventId, details || {});
+    }
+}
+
+export class ToolbarUI extends UIEventTarget {
+    constructor() {
+        super();
+        this.notebookToolbar = new NotebookToolbarUI().setEventParent(this);
+        this.cellToolbar = new CellToolbarUI().setEventParent(this);
+        this.codeToolbar = new CodeToolbarUI().setEventParent(this);
+        this.textToolbar = new TextToolbarUI().setEventParent(this);
+        this.settingsToolbar = new SettingsToolbarUI().setEventParent(this);
+
+        this.el = div(['toolbar-container'], [
+            this.notebookToolbar.el,
+            this.cellToolbar.el,
+            this.codeToolbar.el,
+            this.textToolbar.el,
+            this.settingsToolbar.el,
+        ])
+    }
+
+    onContextChanged() {
+        if (Cell.currentFocus instanceof TextCell) {
+            this.el.classList.remove('editing-code');
+            this.el.classList.add('editing-text');
+        } else if (Cell.currentFocus instanceof CodeCell) {
+            this.el.classList.remove('editing-text');
+            this.el.classList.add('editing-code');
+        }
+    }
+}
+
+function toolbarElem(name, buttonGroups) {
+    const contents = [h3([], [name])].concat(
+        buttonGroups.map(group => {
+            if (group instanceof Array) {
+                return div(["tool-group"], group)
+            } else {
+                return div(["tool-group"].concat(group.classes), group.elems)
+            }
+        }));
+    return div(["toolbar", name], contents)
+}
+
+// iconButton that doesn't steal focus
+function toolbarButton(classes, title, icon, alt) {
+    return iconButton(classes, title, icon, alt)
+        .listener('mousedown', (evt) => evt.preventDefault())
```

Comment:
I think previously the root toolbar element had a `mousedown` listener to prevent default. Seemed to work OK for preventing focus-stealing across the entire toolbar.

Reply:
Ooh that would be a lot cleaner, I'll make that change. 

I was copying from https://github.com/polynote/polynote/blob/master/polynote-frontend/polynote/toolbar.js#L29 which I think is was doing this previously.

Reply:
Done, much nicer now thanks :)